src/index.js

```diff
@@ -170,224 +172,232 @@ class UglifyJsPlugin {
       const tasks = [];
 
       const { chunkFilter } = this.options;
+      const cacheDir = findCacheDir({ name: 'uglifyjs-webpack-plugin' });
 
-      Array.from(chunks)
-        .filter((chunk) => chunkFilter && chunkFilter(chunk))
-        .reduce((acc, chunk) => acc.concat(chunk.files || []), [])
-        .concat(compilation.additionalChunkAssets || [])
-        .filter(ModuleFilenameHelpers.matchObject.bind(null, this.options))
-        .forEach((file) => {
-          let inputSourceMap;
+      const promise = Promise.all(
+        Array.from(chunks)
+          .filter((chunk) => chunkFilter && chunkFilter(chunk))
+          .reduce((acc, chunk) => acc.concat(chunk.files || []), [])
+          .concat(compilation.additionalChunkAssets || [])
+          .filter(ModuleFilenameHelpers.matchObject.bind(null, this.options))
+          .map(async (file) => {
+            let inputSourceMap;
 
-          const asset = compilation.assets[file];
+            const asset = compilation.assets[file];
 
-          if (processedAssets.has(asset)) {
-            return;
-          }
+            if (processedAssets.has(asset)) {
+              return;
+            }
 
-          try {
-            let input;
+            try {
+              let cacheKey;
+              let cacheHit = false;
+              const input = asset.source();
+
+              if (this.options.cache) {
+                const defaultCacheKeys = {
+                  // eslint-disable-next-line global-require
+                  'uglify-js': uglifyJsPackageJson.version,
+                  // eslint-disable-next-line global-require
+                  'uglifyjs-webpack-plugin': require('../package.json').version,
+                  'uglifyjs-webpack-plugin-options': this.options,
+                  hash: crypto
+                    .createHash('md4')
+                    .update(input)
+                    .digest('hex'),
+                };
+
+                cacheKey = serialize(defaultCacheKeys);
+                cacheHit = !!(await cacache.get.info(cacheDir, cacheKey));
+              }
 
-            if (this.options.sourceMap && asset.sourceAndMap) {
-              const { source, map } = asset.sourceAndMap();
```

Comment:
Apparently sourceAndMap is supposed to be faster than calling `asset.source()` and `asset.map()` separately, but not having to call `asset.map()` if we have the output source map cached is even faster. :)